### DISCUSSION TITLE
Multiple packages using Firebase

### DISCUSSION THREAD

**Discussion by wassil** [invertase/react-native-firebase#4360] [Q&A] [answered]
Hi,
I'm working on a package `react-native-exponea-sdk` and on android we use firebase for push notifications.
We have reports from our users that when they use `react-native-firebase` in combination with our package, the `react-native-firebase` fails to auto-initialize Firebase. The error they are seing is `No Firebase App [DEFAULT] has been created`. They were able to fix it by adding `io.invertase.firebase.app.ReactNativeFirebaseApp.initializeSecondaryApp("[DEFAULT]", this);` into `MainApplication.onCreate` (https://github.com/exponea/exponea-react-native-sdk/issues/3)
If I understand it correctly, the line just initializes firebase. From documentation it seems that should happen automatically though. 

Can anyone help me understand how this package initializes Firebase and why it doesn't work in case there are 2 packages using Firebase? Is this expected behavior? I'm ok telling our users they need to call that function, I'm just not 100% sure it's the right way to solve this.

**Comment by mikehardy** [comment, marked as answer]
Hi there! 

Here's the most basic integration for react-native-firebase, based on our documentation here (https://rnfirebase.io): 

https://github.com/mikehardy/rnfbdemo/blob/fc7e943ca95fe1777300985f31db2a075d2da4a5/make-demo.sh#L23-L35

Seems odd to initialize the '[DEFAULT]' app for Firebase as a secondary app, so I'm not sure that's correct either. 

Firebase initialization on Android is hooked by the system a little cleverly in that it is done as a ContentProvider since those are instantiated prior to Application::onCreate even

https://github.com/invertase/react-native-firebase/blob/0a80692359e7c0d08fc66def05a70c7167e7c8ca/packages/app/android/src/reactnative/AndroidManifest.xml#L11-L15

similarly, where the default app should be first initialized:

https://github.com/firebase/firebase-android-sdk/blob/e65f4876ff61b5791986c249b21cb094c37bd29e/firebase-common/src/main/java/com/google/firebase/provider/FirebaseInitProvider.java#L48-L57

why is that not working with your library? a mystery unfortunately but it generates info logs. I suspect logcat would be informative

**Reply by wassil** [reply]
Hi Mike,
thanks for the great answer, it really helped me figure this one out!
What happened is that our native android SDK disables the default firebase init provider and initializes firebase manually - which doesn't actually sound like a smart idea ;) 
Knowing you rely on init providers was the key.

**Reply by mikehardy** [reply]
Glad that helped. This is unrelated but if you are digging into this area, one thing I just learned recently is that while this (using ContentProviders to step in front of Application::onCreate) is clever, did you know it is possible for your Application to start and not actually have *either* the ContentProvider *or* Application::onCreate called? It's true - it happens during google play services app backup / restore and was the cause of some low-frequency but very puzzling crashes in a separate project. Android is infinitely surprising: https://github.com/ankidroid/Anki-Android/pull/7631
